[serve] Enrich serve status API

zcin opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Add 2 more statuses to DeploymentStatus:
* UPSCALING
* DOWNSCALING

Add `DeploymentStatusTrigger` to a deployment's state:
* UNSPECIFIED
* DEPLOY
* CONFIG_UPDATE
* UPSCALE_COMPLETED
* DOWNSCALE_COMPLETED
* AUTOSCALE
* DELETE

State diagram:
<img width="1386" alt="Screenshot 2023-11-17 at 9 15 10 AM" src="https://github.com/ray-project/ray/assets/15851518/72a9d4e7-0a18-4094-9f65-ac5b7d6537c6">


Side note on application status:
Application will remain `RUNNING` if all deployments are either HEALTHY or UPSCALING/DOWNSCALING due to autoscaling.


## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

GeneDer commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1368,34 +1401,22 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+        is_scale = target_state.compare_state_excluding_num_replicas(self._target_state)
+        if self._target_state.info and is_scale:
+            new, old = (target_state.num_replicas, self._target_state.num_replicas)
+            scaling_decision = "Upscaling" if new > old else "Downscaling"
```
Also, we should probably use the `DeploymentStatus` eum directly here instead of the string "Upscaling" and "Downscaling"

zcin replied:
New and old shouldn't be equal because for both manual scaling (changing `num_replicas`) and autoscaling, there are defensive checks that return early if nothing has changed. I'll add an assertion here.

GeneDer commented on python/ray/serve/tests/test_deployment_state.py:
```diff
@@ -1788,9 +1981,178 @@ def test_new_version_and_scale_up(mock_deployment_state):
+
+
+@pytest.mark.parametrize("scale_direction", ["up", "down"])
+def test_scale_num_replicas_up(mock_deployment_state_manager_full, scale_direction):
```
Maybe rename the test to drop the suffix "_up" 🙃 same goes to "test_autoscale_up"

GeneDer reviewed (commented):
Some minor comments, but mostly LGTM :)

GeneDer reviewed (approved):
LGTM!

zcin commented:
Based on feedback from Huaiwei and Angelina, `UPSCALE_COMPLETED` and `DOWNCALE_COMPLETED` does not match with the rest of the status triggers. They would prefer `UPSCALE` and `DOWNSCALE` instead; i.e. once Serve successfully scales the number of replicas to the target number, the state will transition from (UPSCALING, AUTOSCALE) -> (HEALTHY, UPSCALE). Please let me know if anyone is against this, otherwise I will go with this naming!

shrekris-anyscale commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1368,34 +1401,27 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+        if self._target_state.info and is_scale:
+            new, old = (target_state.num_replicas, self._target_state.num_replicas)
+            assert new != old
+            scaling_decision = (
+                DeploymentStatus.UPSCALING
+                if new > old
+                else DeploymentStatus.DOWNSCALING
+            )
```
Where's the logic that determines the application status based on these deployment statuses? I think I missed it.

I want to confirm that the application status is `DEPLOYING` while deployments are outside the `min_replicas`/`max_replicas` bounds and then become `RUNNING` as soon as they enter those bounds, even if they're still autoscaling.

edoakes replied:
it's in `application_state`.

I also am not sure if this is handled right now, I don't believe so.

zcin replied:
Yup this is not handled right now, I will add it in a follow up PR! Want to get this one in quickly to unblock folks.

shrekris-anyscale reviewed (commented):
Nice work! I reviewed the code logic, but I haven't taken a look at the tests.

edoakes commented on python/ray/serve/_private/common.py:
```diff
@@ -102,30 +105,53 @@ class DeploymentStatus(str, Enum):
+
+class DeploymentStatusTrigger(str, Enum):
+    UNSPECIFIED = "UNSPECIFIED"
+    DEPLOY = "DEPLOY"
+    CONFIG_UPDATE = "CONFIG_UPDATE"
```
what's the difference between these two?

edoakes replied:
oh is "DEPLOY" a heavyweight config change?

zcin replied:
DEPLOY is when a new deployment is created, CONFIG_UPDATE is when an existing deployment is being updated with new values. (which is why when the target state has no target info yet, [I set the trigger to DEPLOY](https://github.com/ray-project/ray/pull/40951/files#diff-c4c2583a4c2f3a3c87ada6faebd0b2dfef404e165df133a11195be5d65fcb387R1465-R1470).

edoakes replied:
I see

edoakes commented on python/ray/serve/_private/common.py:
```diff
@@ -102,30 +105,53 @@ class DeploymentStatus(str, Enum):
     def debug_string(self):
         return json.dumps(asdict(self), indent=4)
 
+    def update(self, **kwargs):
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
```
can't you just set the fields directly? e.g., `status.name = "whatever"`

zcin replied:
Switched to update because previously we were reconstructing a new `DeploymentStatusInfo` each time we modify the status but sometimes I want to just update one field. I made a method for it because sometimes I want to update multiple fields at the same time so it's cleaner imo to have an update method.

edoakes commented on python/ray/serve/tests/unit/test_application_state.py:
```diff
@@ -236,6 +237,44 @@ def test_unhealthy(self, get_deployments_statuses, mocked_application_state):
+                "c", DeploymentStatus.DOWNSCALING, DeploymentStatusTrigger.CONFIG_UPDATE
+            ),
+        ]
+        status, error_msg = app_state._determine_app_status()
```
here's a case where having a static pure function would make testing more convenient -- pass in whatever is used to determine the status instead of pulling it from instance variables and you could write this test without mocks and/or reaching into internal fields details

edoakes commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -117,6 +118,30 @@ def from_deployment_info(
+        if not other_target_state.info:
+            return False
+
+        return (
+            self.info.replica_config.ray_actor_options
+            == other_target_state.info.replica_config.ray_actor_options
```
you can use
```
return all([
    a == b,
    c == d,
    ...,
])
```

zcin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -421,18 +422,27 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
+                # application is RUNNING
+                num_healthy_or_autoscaling_deployment += 1
+            else:
+                # If deployments are UPDATING or UPSCALING/DOWNSCALING
```
We have an application status `DELETING`, which is dictated by whether the application target state has the `deleting` flag set to true: see [here](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/application_state.py#L421-L422). When the `deleting` flag is set to true, the application state manager no longer derives a status based on the statuses of the deployments. This PR doesn't change that logic.

shrekris-anyscale replied:
Makes sense.

zcin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1774,18 +1818,31 @@ def _check_curr_status(self) -> Tuple[bool, bool]:
+                if self._curr_status_info.status == DeploymentStatus.UPSCALING:
+                    self._curr_status_info.update(
+                        status=DeploymentStatus.HEALTHY,
+                        status_trigger=DeploymentStatusTrigger.UPSCALE,
```
Yup, see my [comment](https://github.com/ray-project/ray/pull/40951#issuecomment-1800177304) above!

shrekris-anyscale replied:
I'm good with that.

zcin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -117,6 +118,30 @@ def from_deployment_info(
+            and self.info.replica_config.max_replicas_per_node
+            == other_target_state.info.replica_config.max_replicas_per_node
+            and self.info.deployment_config.dict(exclude={"num_replicas"})
+            == other_target_state.info.deployment_config.dict(exclude={"num_replicas"})
+            and self.version
+            and self.version == other_target_state.version
```
This is intentional because `version` can theoretically be `None`, and in that case we generate a random version on the spot. In reality I think that code path is only used for v1 deployments because I cleaned up all uses of passing in version=None for v2, but I haven't gotten around to cleaning up the code in deployment state manager yet.

shrekris-anyscale replied:
Got it, could you add a comment mentioning that?

zcin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -421,18 +422,27 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
                 unhealthy_deployment_names.append(deployment_status.name)
-            if deployment_status.status == DeploymentStatus.HEALTHY:
-                num_healthy_deployments += 1
+            elif (
+                deployment_status.status == DeploymentStatus.HEALTHY
+                or deployment_status.status_trigger == DeploymentStatusTrigger.AUTOSCALE
+            ):
+                # If all deployments are HEALTHY or autoscaling, then
+                # application is RUNNING
+                num_healthy_or_autoscaling_deployment += 1
+            else:
+                # If deployments are UPDATING or UPSCALING/DOWNSCALING
+                # with status trigger CONFIG_UPDATE, then application is
+                # still DEPLOYING
+                num_updating_deployments += 1
```
Thought a little bit about it, I think when the new config is applied, the deployment status will be `UPDATING`. Then in the first control loop iteration that follows that, the status will switch to `UPSCALING` because `autoscale()` will adjust the target num replicas to something within the new [min,max] range.

This also made me realize a bigger problem which is if the deployment is in the middle of updating for whatever reason (even if its a heavyweight change), `autoscale()` will change the status to `UPSCALING`. I can probably change the logic to set status to `UPSCALING/DOWNSCALING` only if the current status is already `HEALTHY`.

sihanwang41 replied:
For maintaining all these status change, i think we can add update_trigger_status in the central place (maybe in DeploymentStatusInfo).
So that we can list out all possible combinations (as the graph shows), and raise Error if we mistakenly update some status.

edoakes replied:
@zcin that sounds good. You'll also need to handle the case where you are `UPDATING` and still scaling up, but the replicas are within the bounds of `[min_replicas, max_replicas]`. In that case, the status should flip to `UPSCALING` (or `DOWNSCALING`).

I also like Sihan's suggestion.

zcin replied:
Can you clarify what `update_trigger_status` would do? Right now `UPDATING`, `UPSCALING`, `DOWNSCALING` are set in a centralized place (`set_target_state`), and `HEALTHY` and `UNHEALTHY` must be set in other parts of the deployment state manager just be nature of how we check for them.

In the meantime, I've fixed the behavior that @edoakes brought up and added a test for it.

zcin replied:
@edoakes That's a good point, although I think that's not super straightforward to fix because it's hard to know whether you are still `UPDATING` because of something else (e.g. code version change or `user_config` change) or if you're still `UPDATING` simply because you haven't reached [min, max replicas] yet. Usually I would expect the status to go from `UPDATING` -> `HEALTHY` -> `UPSCALING/DOWNSCALING`, so I'd consider this an edge case. I've created an issue to track it: https://github.com/ray-project/ray/issues/41038

zcin replied:
@sihanwang41 Can you clarify what `update_trigger_status` would do? Right now `UPDATING`, `UPSCALING`, `DOWNSCALING` are set in a centralized place (`set_target_state`), and `HEALTHY` and `UNHEALTHY` must be set in other parts of the deployment state manager because we check for them in different methods.

In the meantime, I've fixed the behavior that @edoakes brought up and added a test for it.

sihanwang41 replied:
oh, we can explicitly list all possible combinations to avoid setting up unexpected status. (e.g.  deployment status "UPSCALING" + trigger status "UPSCALE_COMPLETED")

zcin replied:
I see, that makes sense! I spent some time prototyping something that is more centralized and it seems promising, but would prefer to add it to a follow up PR since it changes some existing implementation details.

zcin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1368,34 +1401,27 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+        is_scale = target_state.compare_state_excluding_num_replicas(self._target_state)
+        if self._target_state.info and is_scale:
+            new, old = (target_state.num_replicas, self._target_state.num_replicas)
+            assert new != old
```
It's always true because `deploy()` returns early if nothing has changed, and `autoscale()` does the same.

shrekris-anyscale replied:
Could you add that as a comment?

zcin commented:
Other than Sihan's comment above which I will address in a follow up PR, I have addressed all comments! @shrekris-anyscale @sihanwang41 @edoakes please take another look!

shrekris-anyscale commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1368,34 +1407,35 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+        if self._target_state.info and is_scale:
+            if self._curr_status_info.status != DeploymentStatus.UPDATING:
+                new, old = (target_state.num_replicas, self._target_state.num_replicas)
+                # New replicas should never be equal to old replicas because both
```
Is this still true, now that the number of replicas can be scaled by `target_capacity`?

E.g. if you apply a new config with the same number of replicas, but a smaller `target_capacity`, then does the deployment transition to `DOWNSCALING`?

zcin replied:
The comment is still true even with `target_capacity` because `target_capacity` does not affect the target state, it is used in the read-only path (e.g. the scaling method reads `target_state.num_replicas * target_capacity`).

If you apply a new config with num replicas the same + a lower `target_capacity`, the status doesn't change. This is a good thing to bring up, I think we should have an offline discussion about what the status should transition to in this case (will not be addressing in this PR).

shrekris-anyscale commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1368,34 +1407,35 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+                self._curr_status_info.update(
+                    status=scaling_decision,
+                    status_trigger=status_trigger,
+                    message=f"{scaling_decision.capitalize()} from {old} to {new} replicas.",  # noqa
```
[Nit] Why is there a `#noqa` here? Can you split the string into multiple lines with parentheses?

shrekris-anyscale commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1368,34 +1407,35 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
-        self._curr_status_info = DeploymentStatusInfo(
-            self.deployment_name, DeploymentStatus.UPDATING
-        )
-        self._replica_constructor_retry_counter = 0
-        self._backoff_time_s = 1
```
Why don't we reset these values inside `_set_target_state()` anymore? Resetting them here is nice because it's easy to forget to reset them explicitly after calling `_set_target_state()`.

zcin replied:
`_set_target_state` is only called in autoscale() and deploy(). We don't want to reset these values in `autoscale()` so I moved it to deploy()

zcin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1774,18 +1832,31 @@ def _check_curr_status(self) -> Tuple[bool, bool]:
             == 0
         ):
-            # Check for deleting.
+            # Check for deleting and a non-zero number of deployments.
```
I just moved the comment from below.

shrekris-anyscale reviewed (approved):
Nice work!

zcin commented on python/ray/serve/_private/common.py:
```diff
@@ -102,30 +105,53 @@ class DeploymentStatus(str, Enum):
     UPDATING = "UPDATING"
     HEALTHY = "HEALTHY"
     UNHEALTHY = "UNHEALTHY"
+    UPSCALING = "UPSCALING"
+    DOWNSCALING = "DOWNSCALING"
```
> Actually, thinking about it a bit more, why do we even need this standalone state, provided that we already have the trigger specifying the intent for us making it clear that deployment is _updating_ b/c of auto-scaling

I think autoscaling should be it's own state(s). Updating means the configuration for a deployment has changed (modified by the user). For that reason, before this PR, the state of a deployment remains `HEALTHY` when a deployment is autoscaling. For more observability, we added explicit autoscaling states.

> Why do we need 2 states, can we just do it with 1 (AUTOSCALING)?

This is also for more observability, to show the direction the deployment is autoscaling in.

alexeykudinkin replied:
State has to be reflective of the _state_ of the entity (application, deployment), while _trigger_ captures what actually provoked the state-transition.

We need to carefully balance these 2: state machine should be concise, representative and orthogonal (for ex, we'd not communicate whether we're up- or down-scaling using state, b/c it is not reflecting the state, but an action taken against it)

For ex, the question we need to answer how are UPDATING and AUTOSCALING states differ from the perspective of the entity itself (rather than external perspective when we factor in the trigger)?

zcin replied:
Synced offline, we'll be keeping the UPSCALING/DOWNSCALING states but making the status trigger more verbose.

zcin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1369,34 +1408,37 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+                # deploy() and autoscale() return early if the new number of replicas
+                # is the same, so _set_target_state should never get an exact copy of
+                # the existing target state.
+                assert new != old
```
New and old should never be the same here. Do you mean that we should leave the comment but remove the assertion?

alexeykudinkin replied:
Is it an appropriate behavior to reject state update in this case or should we just log it as a warning/error instead?

zcin replied:
Synced offline, will be adding a full coverage unit test for the compare num replicas function and leaving assertion as-is.

zcin commented on python/ray/serve/_private/common.py:
```diff
@@ -102,30 +105,53 @@ class DeploymentStatus(str, Enum):
+
+class DeploymentStatusTrigger(str, Enum):
+    UNSPECIFIED = "UNSPECIFIED"
+    DEPLOY = "DEPLOY"
```
cc: @scottsun94 What do you think?

I am not super opinionated on this, but I think it is more important to have the actual status be a gerund (e.g. `deploying`). Not sure what the best semantics are for the status trigger.

scottsun94 replied:
Hmm. But other triggers are also verbs? Like delete, upscale, autoscale? I'm fine with fully using verb or noun as long as it's consistent.

cc @angelinalg for thoughts too

alexeykudinkin replied:
Using verbs as nouns doesn't make sense, it's not a proper english (we need to fix all of them)

angelinalg replied:
Intuitively, `trigger` is a noun, so I expect nouns for the type:
DEPLOYMENT
AUTOSCALE
UPSCALE
DOWNSCALE
DELETION

A gerund is a bit awkward when the status is HEALTHY and "Deployment completed".
I'm using UPSCALE as a noun like UPGRADE, which can be a noun or a verb (even adjective!).

angelinalg replied:
But we shouldn't belabor the point. Consistency trumps.

zcin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1780,9 +1838,11 @@ def _check_curr_status(self, target_num_replicas: int) -> Tuple[bool, bool]:
                     status=DeploymentStatus.UNHEALTHY,
+                    status_trigger=DeploymentStatusTrigger.UNSPECIFIED
+                    if self._curr_status_info.status == DeploymentStatus.HEALTHY
+                    else self._curr_status_info.status_trigger,
```
The trigger will now be set explicitly to the reason (either health check failure or replica startup failure)

alexeykudinkin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -423,18 +424,28 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
+            ):
+                # If all deployments are HEALTHY or autoscaling, then
+                # application is RUNNING
+                num_healthy_or_autoscaling_deployments += 1
```
Let's avoid mixing up healthy and autoscaling deployments, and instead count them separately (keeps code cleaner and easier to reason about)

alexeykudinkin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -423,18 +424,28 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
+                # If deployments are UPDATING or UPSCALING/DOWNSCALING
+                # with status trigger CONFIG_UPDATE, then application is
+                # still DEPLOYING
+                num_updating_deployments += 1
```
What if deployment is UPDATING and status_trigger == AUTOSCALING, how are we planning to count such Deployment?

zcin replied:
That shouldn't be possible. If status trigger is autoscaling, then the updated target state should only have num replicas changed, so status will be set to upscaling/downscaling.

alexeykudinkin replied:
Got it. Why do we need to mix it into this conditional then?

zcin replied:
I've separated it out into a separate case

alexeykudinkin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1369,34 +1410,39 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
-        logger.info(
-            f"Deploying new version of deployment {self.deployment_name}{app_msg}."
+        # Determine whether the updated target state is simply scaling
+        is_scale = target_state.equal_to_state_excluding_num_replicas(
```
nit: `is_scaling`

alexeykudinkin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1369,34 +1410,39 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
+        # If deployment is not in the middle of updating, and this new
+        # target state only changes the number of replicas, then set
+        # status to be UPSCALING/DOWNSCALING
+        if self._target_state.info and is_scale:
```
Why would we mark state as updating if _target_state.info is absent?

zcin replied:
If it is absent, then this is the first time `deploy()` has been called i.e. it's a new deployment.

alexeykudinkin replied:
Thanks for clarifying! Would be great to call out this context in the comment for future readers

alexeykudinkin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1806,18 +1866,47 @@ def _check_curr_status(self, target_num_replicas: int) -> Tuple[bool, bool]:
-                self._curr_status_info = DeploymentStatusInfo(
-                    self.deployment_name, DeploymentStatus.HEALTHY
-                )
+                if self._curr_status_info.status == DeploymentStatus.UPSCALING:
+                    self._curr_status_info.update(
+                        status=DeploymentStatus.HEALTHY,
+                        status_trigger=DeploymentStatusTrigger.UPSCALE_COMPLETED,
+                        message="",
+                    )
+                elif self._curr_status_info.status == DeploymentStatus.DOWNSCALING:
+                    self._curr_status_info.update(
+                        status=DeploymentStatus.HEALTHY,
+                        status_trigger=DeploymentStatusTrigger.DOWNSCALE_COMPLETED,
+                        message="",
+                    )
+                elif (
+                    self._curr_status_info.status == DeploymentStatus.UPDATING
+                    and self._curr_status_info.status_trigger
+                    == DeploymentStatusTrigger.CONFIG_UPDATE_STARTED
+                ):
+                    self._curr_status_info.update(
+                        status=DeploymentStatus.HEALTHY,
+                        status_trigger=DeploymentStatusTrigger.CONFIG_UPDATE_COMPLETED,
+                        message="",
+                    )
+                elif self._curr_status_info.status == DeploymentStatus.UNHEALTHY:
+                    self._curr_status_info.update(
+                        status=DeploymentStatus.HEALTHY,
+                        status_trigger=DeploymentStatusTrigger.UNSPECIFIED,
+                        message="",
+                    )
+                else:
+                    self._curr_status_info.update(
+                        status=DeploymentStatus.HEALTHY, message=""
+                    )
+
```
Let's customize only status_trigger in this branches, and extract common parts

alexeykudinkin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1961,9 +2049,13 @@ def _check_and_update_replicas(self, target_num_replicas: int):
                 if replica.version == self._target_state.version:
-                    self._curr_status_info: DeploymentStatusInfo = DeploymentStatusInfo(
-                        name=self.deployment_name,
+                    # If the deployment transitioned from HEALTHY to
+                    # UNHEALTHY, then there isn't an explicit status
+                    # trigger since it wasn't because a new deployment or
+                    # an update failed.
```
This comment is misleading -- failing a health-check is an explicit status trigger

zcin replied:
Outdated comment, updated!

alexeykudinkin commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -2112,9 +2199,9 @@ def update(
-                name=self.deployment_name,
+            self._curr_status_info.update(
                 status=DeploymentStatus.UNHEALTHY,
+                status_trigger=DeploymentStatusTrigger.UNSPECIFIED,
```
We should distinguish b/w UNSPECIFIED (aka unset) vs INTERNAL_FAILURE (aka generic, unknown precisely)

zcin replied:
Should we expose users to a trigger named INTERNAL_FAILURE?

alexeykudinkin replied:
Not sure fully understand angle of your question: 
 - We do expose INTERNAL_FAILURE as status in gRPC/HTTP handlers
 - We can name it differently as well, but 
 - We'd make sure we distinguish it from UNSPECIFIED

zcin replied:
I see, I'm OK with adding this as a trigger. I've named it `INTERNAL_ERROR` since `INTERNAL_FAILURE` is slightly inconsistent with triggers like `HEALTH_CHECK_FAILED`.

alexeykudinkin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -424,17 +425,23 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
+                deployment_status.status_trigger == DeploymentStatusTrigger.AUTOSCALING
+            ):
+                num_autoscaling_deployments += 1
+            else:
```
This conditioning is brittle -- we add a new state and this will silently break.

Instead, do it explicit:

```
if HEALTHY:
  ...
elif UNHEALTHY:
  ...
elif UPDATING:
  ...
else:
  raise RuntimeError
```

zcin replied:
Updated!

zcin commented:
@architkulkarni This PR is ready for merge!
